services/grpc/grpc.js

```diff
@@ -0,0 +1,230 @@
+  }
+
+  /**
+   * Log successful disconnect.
+   */
+  async onAfterDisconnect() {
+    grpcLog.info('Disconnected from lnd gRPC service')
+  }
+
+  /**
+   * Rejig connections as needed before activating the wallet unlocker service.
+   */
+  async onBeforeActivateWalletUnlocker() {
+    if (this.services.Lightning.can('disconnect')) {
+      await this.services.Lightning.disconnect()
+    }
+    if (this.services.WalletUnlocker.can('connect')) {
+      await this.services.WalletUnlocker.connect()
+    }
+    await this.services.WalletUnlocker.activate()
+  }
+
+  /**
+   * Rejig connections as needed before activating the lightning service.
+   */
+  async onBeforeActivateLightning() {
+    if (this.services.WalletUnlocker.can('disconnect')) {
+      await this.services.WalletUnlocker.disconnect()
+    }
+    if (this.services.Lightning.can('connect')) {
+      await this.services.Lightning.connect()
+    }
+    await this.services.Lightning.activate()
+  }
+
+  // ------------------------------------
+  // Helpers
+  // ------------------------------------
+
+  /**
+   * Connect all services.
+   */
+  async connectAll() {
+    await Promise.all(
+      Object.keys(this.services).map(serviceName => {
+        const service = this.services[serviceName]
+        if (service && service.can('connect')) {
+          return service.connect()
+        }
+      })
+    )
+  }
+
+  /**
+   * Disconnect all services.
+   */
+  async disconnectAll() {
+    await Promise.all(
+      Object.keys(this.services).map(serviceName => {
+        const service = this.services[serviceName]
+        if (service && service.can('disconnect')) {
+          return service.disconnect()
+        }
+      })
+    )
+  }
+
+  /**
+   * Probe to determine what sate lnd is in.
+   */
+  async determineWalletState() {
+    grpcLog.info('Attempting to determine wallet state')
+    try {
+      await this.services.WalletUnlocker.unlockWallet('null')
+    } catch (e) {
+      switch (e.code) {
+        /*
+          `UNIMPLEMENTED` indicates that the requested operation is not implemented or not supported/enabled in the
+           service. This implies that the wallet is already unlocked, since the WalletUnlocker service is not active.
+           See https://github.com/grpc/grpc-node/blob/master/packages/grpc-native-core/src/constants.js#L129
+         */
+        case status.UNIMPLEMENTED:
+          return WALLET_STATE_ACTIVE
+
+        /**
+          `UNKNOWN` indicates that unlockWallet was called without an argument which is invalid.
+          This implies that the wallet is waiting to be unlocked.
+        */
+        case status.UNKNOWN:
+          return WALLET_STATE_LOCKED
+
+        /**
+          Bubble all other errors back to the caller and abort the connectiopn attempt.
+          Disconnect all services.
+        */
+        default:
+          grpcLog.warn('Unable to connect to lnd service', e)
+          await this.disconnectAll()
```

Comment:
This theoretically could cause exception inside this catch, and `disconnectAll` doesn't handle exceptions itself, so this will result in an unhandled exception